### ISSUE TITLE
Crashing when moving form

### ISSUE THREAD

**Issue by EternityX** [KN4CK3R/OSHGui#9]
I can accurately reproduce this crash every time by moving the form around for 3-5 seconds.

Crash occurs on https://github.com/KN4CK3R/OSHGui/blob/master/Controls/Control.cpp#L513

![image](https://i.imgur.com/mRcVbx3.png)

geometry_ virtual table is nullptr?
![image](https://i.imgur.com/UbuyY1x.png)

I'm not quite sure how to fix this issue. I can provide more information if needed.

Edit: This might be an issue that is caused by some of my changes to OSHGui, I'll continue looking into it

**Comment by KN4CK3R** [comment]
For me it looks like the current "this" objects memory was released (use after free). Let me know if this is an bug in my code. I have never experienced this in all the years.

**Comment by EternityX** [comment]
I've reverted all my changes to the original code found in this github repo and the issue still occurs.

Could there be something wrong with my initialization?

renderer initialization:
```c
void c_renderer::init( IDirect3DDevice9 *device ) {
     m_renderer = std::make_unique< OSHGui::Drawing::Direct3D9Renderer >( device );

     // initialize oshgui with our renderer.
     OSHGui::Application::Initialize( std::move( m_renderer ) );

     // setup fonts here.

     OSHGui::Application::Instance( ).SetDefaultFont( m_fonts.at( FONT_VERDANA_7PX ) );
}
```

form creation:
```c
void c_menu::init( ) {
    m_form = std::shared_ptr < OSHGui::Form >( new c_main_form( ) );
    m_form->SetFont( OSHGui::Application::Instance( ).GetDefaultFont( ) );

    OSHGui::Application::Instance( ).Run( m_form );
    OSHGui::Application::Instance( ).Enable( );
}
```

tab control + pages:
```c
void c_main_form::init_tabs( ) {
    m_tab_control = new OSHGui::TabControl( );

    // create pages.
    for( int i = 0; i < PAGE_MAX; ++i )
        m_pages.push_back( std::make_shared< OSHGui::TabPage >( ) );

    m_tab_control->SetFont( OSHGui::Application::Instance( ).GetDefaultFont( ) );

    // set page titles.
    m_pages.at( PAGE_RAGEBOT )->SetText( "Ragebot" );
    m_pages.at( PAGE_VISUALS )->SetText( "Visuals" );
    m_pages.at( PAGE_MISC )->SetText( "Miscellaneous" );
    m_pages.at( PAGE_CONFIG )->SetText( "Configuration" );
    m_pages.at( PAGE_PLAYERLIST )->SetText( "Skinchanger" );

    // set tab style.
    m_tab_control->SetFont( OSHGui::Application::Instance( ).GetDefaultFont( ) );
    m_tab_control->SetSize( 576, 380 );
    m_tab_control->SetBackColor( OSHGui::Drawing::Color::FromRGB( 32, 32, 38 ) );
    m_tab_control->SetLocation( 6, -15 );
    //m_tab_control->SetButtonWidth( 116 );

    // add all pages to tab control.
    for( auto &item : m_pages )
        m_tab_control->AddTabPage( item.get( ) );

    this->AddControl( m_tab_control );
}
```

control creation (if I add more than 10-15~ checkboxes it crashes much faster):
```c
auto checkbox = new OSHGui::CheckBox( );
checkbox->SetText( "test" );
m_pages.at( PAGE_RAGEBOT )->AddControl( checkbox );
```

Interestingly, if I add:
```c
if( geometry_.unique( ) )
    geometry_->SetTranslation(Drawing::Vector(absoluteLocation_.X, absoluteLocation_.Y, 0.0f));
```
I no longer get the crash, but of course this messes up the movement of the form.

**Comment by KN4CK3R** [comment]
Could you attach your project here? If it should not be public you can send it to admin [ at ] oldschoolhack.me

**Comment by EternityX** [comment]
All related code can be found @ https://github.com/EternityX/DEADCELL-CSGO

some files of interest are:
https://github.com/EternityX/DEADCELL-CSGO/blob/master/deadcell/oshgui/renderer.cpp
https://github.com/EternityX/DEADCELL-CSGO/blob/master/csgo/menu/menu.cpp

and here's an issue related to the crash https://github.com/EternityX/DEADCELL-CSGO/issues/23

Thank you very much for your support 😄

**Comment by KN4CK3R** [comment]
I will have a look at it tomorrow. Because of all the screenshots in the other issue I assume the stack gets corrupted somewhere and it's not directly related to the gui but we will see. Is the project game ready? Just compile and inject into CSGO?

**Comment by EternityX** [comment]
> I will have a look at it tomorrow. Because of all the screenshots in the other issue I assume the stack gets corrupted somewhere and it's not directly related to the gui but we will see. Is the project game ready? Just compile and inject into CSGO?

Yes the project should compile and inject fine.
I generally inject at main menu, load into map, and open menu and drag menu around until it eventually crashes.

**Comment by ghost** [comment]
Hi, when we turn off multi core rendering the crash no longer occurs, but my question is why does it crash with multi core rendering on. Multi core rendering is a video setting in csgo btw.

**Comment by KN4CK3R** [comment]
I had the same problem some years ago when I created CSGO hack. I simply implemented a `class CSGORenderer : public OSHGui::Drawing::Renderer` which uses `IMatSystemSurface` for all rendering. Because OSHGui uses textures I couldn't draw the gui with the new renderer but the esp stuff in a PaintTraverse hook. If I remember correctly you can't use engine methods like `TraceRay` in the multi core DX environment because they are not thread safe. The calls corrupted the stack and you end up with the strange errors you are seeing.

For a test you could remove the lines

https://github.com/EternityX/DEADCELL-CSGO/blob/a8266f2477ffbaf39d68b5b5e096d4d19f892e11/csgo/hooking/hooks/present.cpp#L22-L35

and test if it still crashes with multi core rendering. If it stops crashing you have your answer. In this case this isn't a OSHGui problem because every DX drawing mixed with engine calls will fail.

Edit: A quote from the release notes from 2012-11-25

> Q: ESP flickers / OSHFusion crashes often in Source Engine games.
> A: Deactivate Multicore Rendering in advanced graphics options.

**Comment by ghost** [comment]
Yeah but the weird thing, it only happens when moving menu in game, turning off multi core rendering for people fixed the flicker issue and the crash, I still don’t see why it crashes while in game, could it be that we use engine functions in present thread?, we’re not using anything related to traceray tho.

**Comment by KN4CK3R** [comment]
You are using TraceRay in the nade prediction method:
https://github.com/EternityX/DEADCELL-CSGO/blob/a8266f2477ffbaf39d68b5b5e096d4d19f892e11/csgo/features/nade_pred/nade_pred.cpp#L67-L70

Edit: My bad, the trace method only gets called from CreateMove here.

How about switching to drawing in PaintTraverse? I can upload my renderer classes if you need them. You would draw the gui in DX Present and esp, ... in PaintTraverse. You are already hooking this method so it should not be a big restructure.

**Comment by EternityX** [comment]
I've already commented out every single hook besides present and reset and it still crashes with multicore rendering. 

Drawing in Paint Traverse would actually fix some other issues we have had, so it would be really helpful if you did upload them.

**Comment by ghost** [comment]
It would be very helpful if you could provide your renderer classes, it would help us resolve a couple problems we have with the cheat.

**Comment by KN4CK3R** [comment]
Here is the code:
https://gist.github.com/KN4CK3R/59ff579afd2ed236730757476e41239a

**Comment by IcePixelx** [comment]
Weirdly I have the same problem but it only occurs when compiling in debug. In release its fine.
And it seems to be somewhat csgo related since this problem only occurs for me ingame.

Gonna post some code maybe it helps someone lol.

```
HRESULT __stdcall Hooks::hkEndScene(IDirect3DDevice9* device)
{
	static auto oEndScene = d3ddevice_hook->get_original<EndSceneFn>(Index::EndScene);

	SaveState(device);

	/*
	Steamoverlay stuff here
	*/

	static bool init = [&]() // & is allowing us to capture all enclosing variables
	{
		g_OSHGuiRenderer.Init(device);
		g_Menu.Init();
		g_InputManager.Init("Valve001", g_OSHGuiRenderer.GetInstance());
		return true;
	} ();

	[device]() 
	{
		if (!g_DX9Renderer.IsInitialized())
		{
			g_DX9Renderer.InitDeviceObjects(device);
			g_DX9Renderer.SetInitialized(true);
		}
		else
		{
			g_DX9Renderer.SetupRenderStates();
			g_DX9Renderer.RenderDrawList(); 
		}
	}();

	g_OSHGuiRenderer.Draw();

	RestoreState(device);
	return oEndScene(device);
}

**Comment by ghost** [comment]
@KN4CK3R, I have no crash at all when moving forms.

But having many elements (even just a few), I just can't move the form smoothly, it is **very sluggish**.

I am using many tabs and testing on an empty one everything is alright when moving the main form, but on a tab with a few elements, again, moving is sluggish.

I actually don't know if there is a way to optimize that.
